src/hooks/useViewportRows.ts

```diff
@@ -0,0 +1,102 @@
+import { useMemo } from 'react';
+
+import { GroupRow, GroupByDictionary, Dictionary } from '../types';
+import { getVerticalRangeToRender } from '../utils';
+
+interface ViewportRowsArgs<R, SR> {
+  rawRows: readonly R[];
+  rowHeight: number;
+  clientHeight: number;
+  scrollTop: number;
+  groupBy: readonly string[];
+  rowGrouper?: (rows: readonly R[], columnKey: string) => Dictionary<readonly R[]>;
+  expandedGroupIds?: ReadonlySet<unknown>;
+}
+
+export function useViewportRows<R, SR>({
```

Comment:
Can we move `getViewportRows` in this hook? 🤔

Reply:
I am not too sure but I do want to do some cleanup as `DataGrid` code is getting complex. Let's think about it and improve it in a followup PR